src/ast/json.rs

```diff
@@ -437,6 +438,98 @@ impl<'arena> Ast<'arena> for Json<'arena> {
     fn is_valid_root(&self, _node_type: Class) -> bool {
         true
     }
+
+    fn add_to_graph(
+        &'arena self,
+        digraph_edges: &mut String,
+        hmap_nodes: &mut HashMap<usize, String>,
+    ) {
```

Comment:
This function doesn't need to be implemented specifically for each `Ast` - everything except the node label is the same for every node of every possible Ast type.

You can use `some_node.children()` to get a slice of children of a given node (which will be empty for nodes like `true` and `false`).

I think something like this would be better:
```rust
// ast/mod.rs
trait Ast {
    /// The name of this node as should be displayed in the DAG debug graph
    fn debug_name(&self) -> String;
}

// editor/dag.rs
impl Dag<'arena, Node: Ast<'arena>> {
    // --- snip ---

    /// Generate `.dot` code of the internal state of the `Dag`
    pub fn to_dot_code(&self) -> String {
        // This replaces `Ast::add_to_graph`, so we don't duplicate functionality
        fn add_to_graph(
            node: &'arena Node,
            digraph_edges: &mut String,
            hmap_nodes: &mut HashMap<usize, String>,
        ) {
            let name = node as *const Node as usize;
            hmap.nodes.entry(name).or_insert(format!(
                "node{} [label=\"{}\"]\n",
                name,
                node.debug_name()
            ));
        }
        // Basically the same code as you already have
    }
}
```

Reply:
Yes, I wondered if there is a better way to do this. I think your review is exactly what I was looking for. :)

Reply:
I can't define the `add_to_graph` inside `to_dot_code`. I got an error: 'can't use generic parameters from outer function'.
So I made the `add_to_graph` as an private associated function.

Reply:
I have fixed the problem by adding lifetime and generic into the function signature. 
```rust
        fn add_to_graph<'arena, Node: ast<'arena>> (
            node: &'arena Node,
            digraph_edges: &mut String,
            hmap_nodes: &mut HashMap<usize, String>,
        ) {
            let name = node as *const Node as usize;
            hmap.nodes.entry(name).or_insert(format!(
                "node{} [label=\"{}\"]\n",
                name,
                node.debug_name()
            ));
        }
```
I have a question that since this inner function and the Dag have same lifetime `'arena` notation, their lifetime have similar length.  The lifetime doesn't be influenced by being a inner function.

Reply:
Aha so in Rust, private functions are simply stand-alone functions that can only be accessed within the scope they're defined. So you need to specify the parameters again because functions can't capture any local state (we would use a closure to do something like that, but there's no way to do recursion inside a closure and we need to do that).

This is very different to how Python works, which is probably why the behaviour comes as a surprise. By the way, I've been playing with local functions in python and they're actually incredibly broken - they capture variables from the scope in which they are **called**, now when they're **defined**, as I expected.

So this works just fine:
```python
def main():
  def print_x():
    print(x) # x isn't defined when the function is created... 

  x = 3
  print_x() # ... but this prints 3
```